### PR TITLE
[2/n] SFTDataset: refactor slimorca and message converters

### PR DESCRIPTION
**Pull request by RdoubleA**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [x] update tests and/or documentation
- [ ] other (please add here)

As discussed in the RFC in #1186, we will merge instruct and chat datasets to the following unified pipeline that can better support multimodal:
1. `message_transform` to create `List[Message]` from dataset with full flexibility on columns, ad-hoc modifications, etc. For multimodal, additionally images are loaded from the path
2. `prompt_template` as a optional way to add structured text around specific roles in the list of messages
3. `model_transform` that takes the list of messages and tokenizes it. For multimodal, it will additionally apply model-specific image transforms to the images associated with the sample

**For ease of review, we will stage this as multiple moderate-sized PRs**. This PR updates `slimorca_dataset` to use `SFTDataset` and refactors the message converters `get_sharegpt_messages` and `get_openai_messages` to their transform analogues: `ShareGPTToMessages` and `JSONToMessages`

Also renames `_finetune.py` to `_sft.py`

Previous PR: #1234 

#### Test plan
- [x] update live docs and check rendering
- [x] unit tests for `ShareGPTToMessages`, `JSONToMessages`
- [x] updated unit test for `slimorca_dataset`
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [x] compare loss curves of slimorca to original versions on main
image

### CODE REVIEW

File: torchtune/datasets/_slimorca.py [hunk 1]
```diff
@@ -4,53 +4,54 @@
...
-from torchtune.datasets._chat import chat_dataset, ChatDataset
+from torchtune.data import PromptTemplate, ShareGPTToMessages
 
-from torchtune.modules.tokenizers import ModelTokenizer
+from torchtune.datasets._finetune import SFTDataset
+from torchtune.datasets._packed import PackedDataset
+from torchtune.modules.transforms import Transform
 
 
 def slimorca_dataset(
```
Comment (joecummings) [inline torchtune/datasets/_slimorca.py:16]: This whole thing is so nice.

File: torchtune/datasets/_slimorca.py [hunk 2]
```diff
@@ -4,53 +4,54 @@
...
         packed (bool): Whether or not to pack the dataset to ``max_seq_len`` prior to training. Default is False.
         split (str): ``split`` argument for ``datasets.load_dataset``. You can use this argument to load a subset
             of a given split, e.g. ``split="train[:10%]"``. Default is "train".
 
-    Raises:
-        ValueError: If `max_seq_len` is less than 4.
-
     Returns:
-        ChatDataset: dataset configured with SlimOrca source data and Llama2 chat template
+        Union[SFTDataset, PackedDataset]: dataset configured with SlimOrca source data
```
Comment (joecummings) [inline torchtune/datasets/_slimorca.py:54]: Can we use the `|` operator yet for this? Or is that only available in Python 3.11?

Comment (joecummings) [reply torchtune/datasets/_slimorca.py:54]: Like `SFTDataset | PackedDataset`

Comment (RdoubleA) [reply torchtune/datasets/_slimorca.py:54]: 3.10 and above, and I think we test 3.9?

**Comment by SalmanMohammadi** [comment]
Rafi shows me what a 10x engineer looks like, this is fresh clean

File: docs/source/api_ref_data.rst [hunk 3]
```diff
@@ -53,6 +53,19 @@ Converts data from common JSON formats into a torchtune :class:`Message`.
...
+
+.. autosummary::
+    :toctree: generated/
+    :nosignatures:
+
+    InputOutputToMessages
+    ShareGPTToMessages
+    JSONToMessages
+
 Helper funcs
```
Comment (SalmanMohammadi) [inline docs/source/api_ref_data.rst:69]: nit: Helper Functions??

File: torchtune/datasets/_slimorca.py (hunk 2 above)
Comment (joecummings) [reply torchtune/datasets/_slimorca.py:54]: Grrr okay. Soon, soon.

File: torchtune/datasets/_slimorca.py [hunk 4]
```diff
@@ -4,53 +4,54 @@
...
     Args:
-        tokenizer (ModelTokenizer): Tokenizer used by the model that implements the ``tokenize_messages`` method.
-        source (str): path string of dataset, anything supported by Hugging Face's `load_dataset`.
-        chat_format (Optional[str]): name of template used to format the chat. See the description
-            in :class:`~torchtune.datasets.ChatDataset` for more details. Default: None
-        max_seq_len (int): Maximum number of tokens in the returned input and label token id lists.
-            This value needs to be at least 4 though it is generally set to max sequence length accepted by the model.
-            Default is 1024.
+        model_transform (Transform): model specific transform to convert a list of messages
+            output by the dataset to tokens. This will always be a :class:`~torchtune.modules.tokenizers.ModelTokenizer`.
```
Comment (ebsmothers) [inline torchtune/datasets/_slimorca.py:38]: I commented this on the other PR (after it landed), but can you clarify why this always has to be a `ModelTokenizer`? And if that's the case, why don't we type it as such?

**Review by ebsmothers — APPROVED** [review]
One comment, otherwise looks great

File: torchtune/datasets/_slimorca.py (hunk 4 above)
Comment (RdoubleA) [reply torchtune/datasets/_slimorca.py:38]: Discussed offline, keeping this as Transform to maintain consistency across SFTDataset, text dataset builders, multimodal dataset builders

**Review by joecummings — APPROVED** [review]

**Pull request merged**